### PR TITLE
chore: test Flow for edge cases cause by misconfigs

### PR DESCRIPTION
**Pull request by dylanhitt**
Per #322

The following is to test provide a bit easier of test bed for testing edge cases do to server misconfigs or poor user input. The test here will pass, but they bring up some questions:

1. When ErrorHandler is overridden there is a possibility it can return `nil` causing golang encode to eventually land on `null` (I will probably expand this to more return types but wanted to get the conversation going). What should we do here? Is the user setting the return type to nil to say it's fine and we should really respond with like `NoContent` or is just saying 500 with empty/null response good enough. I personally think if the user is setting their error to nil they're doing something on purpose 🤷 
2. In the case of transformOut error when it's receiver is a value and not a pointer. We throw errors.New(""). ErrorStrings are not marshalable. In this case I think we should wrap this error with `fuego.InternalServerError` since transform is supported only on structs with the implementation we can assume a json/xml/x-yaml response is fine depending on the user request.

### CODE REVIEW

**Comment by EwenQuim** [comment]
Good idea. I'd say it's user's will...

> 1. When ErrorHandler is overridden there is a possibility it can return nil causing golang encode to eventually land on null (I will probably expand this to more return types but wanted to get the conversation going). What should we do here? Is the user setting the return type to nil to say it's fine and we should really respond with like NoContent or is just saying 500 with empty/null response good enough. I personally think if the user is setting their error to nil they're doing something on purpose 🤷

Yeah clearly that's user defined on purpose so he knows what he's doing. Thanks for the test, it'll help not break future changes.

> 2. In the case of transformOut error when it's receiver is a value and not a pointer. We throw errors.New(""). ErrorStrings are not marshalable. In this case I think we should wrap this error with fuego.InternalServerError since transform is supported only on structs with the implementation we can assume a json/xml/x-yaml response is fine depending on the user request.

> when it's receiver is a value and not a pointer

It's unfortunate but that's how Go works, when using a value receiver, the struct will not be changed! So Fuego must not do magic here. User expects Go-like behavior. But maybe we can add a warning!

> We throw errors.New(""). ErrorStrings are not marshalable.

Mmmh I see. But same, user does not expect a surprise `fuego.InternalServerError` to appear. I'd say it's better to return a string containing the error?

File: serve_test.go [hunk 1]
```diff
@@ -561,3 +578,39 @@ func newTLSTestHelper() (*tlsTestHelper, error) {
...
+	t.Run("base", func(t *testing.T) {
+		e := NewEngine()
+		w := httptest.NewRecorder()
+		ctx := newTestCtx(w)
+		Flow(e, ctx, testController)
+		assert.Equal(t, http.StatusOK, w.Code)
+		assert.Equal(t, crlf(`{"ans":"Hello World"}`), w.Body.String())
+	})
+	t.Run("with nil return in ErrorHandler", func(t *testing.T) {
+		e := NewEngine(WithErrorHandler(func(err error) error { return nil }))
```
Comment (ccoVeille) [inline serve_test.go:601]: What about a test with


```suggestion
		e := NewEngine(WithErrorHandler(nil)
```

Comment (EwenQuim) [reply serve_test.go:601]: It'll throw, it's already tested elsewhere ;)

**Comment by dylanhitt** [comment]
> Mmmh I see. But same, user does not expect a surprise fuego.InternalServerError to appear. I'd say it's better to return a string containing the error?

Let me see how that would look.

**Comment by dylanhitt** [comment]
> Mmmh I see. But same, user does not expect a surprise fuego.InternalServerError to appear. I'd say it's better to return a string containing the error?

Ehhh there doesn't really seem to be a nice way to do this. Unless we create some internal fuego error type or something, but even then it's not great. Will keep trying...

**Comment by dylanhitt** [comment]
We could override the marshaler to handle this case I guess.

**Comment by EwenQuim** [comment]
MMmh I don't want Fuego to become a "magic" framework!

**Comment by dylanhitt** [comment]
Hence why I personally like fuego.InternalServerError. In reality that's what is happening. Something on the server is erroring the administrator needs to handle it. In the case of text/plain it will just return the `.Error()` of our internal server error.

**Comment by dylanhitt** [comment]
This also just uncovered that we've been doing this on `x-yaml` 😅 

```golang
// SendYAMLError sends a YAML error response.
// If the error implements ErrorWithStatus, the status code will be set.
func SendYAMLError(w http.ResponseWriter, _ *http.Request, err error) {
	status := http.StatusInternalServerError
	var errorStatus ErrorWithStatus
	if errors.As(err, &errorStatus) {
		status = errorStatus.StatusCode()
	}

	w.WriteHeader(status)
	_ = SendYAML(w, nil, err.Error())
}
```

the `err.Error()`...

File: serve_test.go [hunk 2]
```diff
@@ -561,3 +579,70 @@ func newTLSTestHelper() (*tlsTestHelper, error) {
...
+				accept:           "application/xml",
+				expectedResponse: "<HTTPError><title>Internal Server Error</title><status>500</status></HTTPError>",
+			},
+			{
+				accept:           "text/html",
+				expectedResponse: "500 Internal Server Error",
+			},
+			{
+				accept:           "application/x-yaml",
+				expectedResponse: "500 Internal Server Error\n",
```
Comment (ccoVeille) [inline serve_test.go:629]: This one doesn't use crlf method

Comment (dylanhitt) [reply serve_test.go:629]: See #339 as for why or the comment just above 🙃

Comment (dylanhitt) [reply serve_test.go:629]: ah i guess it doesn't affect it. updating

File: serialization.go [hunk 3]
```diff
@@ -58,7 +58,11 @@ func transformOut[T any](ctx context.Context, ans T) (T, error) {
 
 	_, ok := any(ans).(OutTransformer)
 	if ok {
-		return ans, errors.New("OutTransformer must be implemented by a POINTER RECEIVER. Please read the [OutTransformer] documentation")
+		err := errors.New("OutTransformer must be implemented by a POINTER RECEIVER. Please read the [OutTransformer] documentation")
```
Comment (ccoVeille) [inline serialization.go:61]: I'm unsure to understand the formatting of OutTransformer here

I mean it's an error message, not a godoc or a Markdown file

Am I missing something?

**Review by ccoVeille — APPROVED** [review]

File: serialization.go (hunk 3 above)
Comment (dylanhitt) [reply serialization.go:61]: Yeah the only place this should really manifest is in a log server side. Maybe a good time to rethink it's shape.

**Comment by dylanhitt** [comment]
Hey @EwenQuim are you call with this merging?

You had some concerns about switching the error to to our `InternalServerError`. I personally think the return on this is the best we can do. I think it's good to not expose the client to any information.

**Review by EwenQuim — APPROVED** [review]
Good ideas

**Comment by EwenQuim** [comment]
I let you merge, thanks for this PR! A good cleanup
